pkg/plugins/http/sender.go

```diff
@@ -120,18 +128,236 @@ func NewSender(tid tenant.Id, plugin string, name string, config interface{}, se
+		if err != nil {
+			return "", 0, err
+		}
+		req.Header.Add("Authorization", token)
+	} else if s.config.Auth.Type == HTTP_AUTH_TYPE_OAUTH {
+		return "", 0, errors.New("oauth auth not supported")
+	} else if s.config.Auth.Type == HTTP_AUTH_TYPE_OAUTH2 {
+		var ok bool
+		s.RLock()
+		client, ok = s.clients[HTTP_AUTH_TYPE_OAUTH2+"-"+url]
+		s.RUnlock()
+		if !ok {
+			conf := &clientcredentials.Config{
+				ClientID:     s.secrets.Secret(ctx, s.config.Auth.ClientID),
+				ClientSecret: s.secrets.Secret(ctx, s.config.Auth.ClientSecret),
+				TokenURL:     s.secrets.Secret(ctx, s.config.Auth.TokenURL),
+				Scopes:       s.config.Auth.Scopes,
+			}
+			if s.config.Auth.GrantType != "" {
+				conf.EndpointParams = nurl.Values{"grant_type": []string{s.config.Auth.GrantType}}
+			}
+			client = conf.Client(context.Background())
+			s.Lock()
+			s.clients[HTTP_AUTH_TYPE_OAUTH2+"-"+url] = client
+			s.Unlock()
+		}
+	} else if s.config.Auth.Type == "" {
+		var ok bool
+		s.RLock()
+		client, ok = s.clients[HTTP_AUTH_TYPE_BASIC]
+		s.RUnlock()
+		if !ok {
+			client = s.initHttpTransportWithDialer()
+			s.Lock()
+			s.clients[HTTP_AUTH_TYPE_BASIC] = client
+			s.Unlock()
+		}
+	} else {
+		return "", 0, errors.New("invalid auth type " + s.config.Auth.Type)
+	}
+	if client == nil {
+		return "", 0, errors.New("no client found for auth type " + s.config.Auth.Type)
+	}
+	// send request
+	start := time.Now()
+	resp, err := client.Do(req)
+	s.eventSendOutTime.Record(evt.Context(), time.Since(start).Milliseconds())
+	s.eventBytesCounter.Add(evt.Context(), int64(len(payload)))
+	if err != nil {
+		log.Ctx(evt.Context()).Error().Str("op", "filter").Str("filterType", "ws").Str("name", s.Name()).Msg("Request error: " + err.Error())
+		return "", 0, err
+	}
+	// read response
+	var body []byte
+	if resp != nil && resp.Body != nil {
+		var readErr error
+		body, readErr = io.ReadAll(resp.Body)
+		if readErr != nil {
+			log.Ctx(evt.Context()).Error().Str("op", "filter").Str("filterType", "ws").Str("name", s.Name()).Msg("Read error: " + readErr.Error())
+			return "", resp.StatusCode, readErr
+		}
+		resp.Body.Close()
+		if body == nil {
+			return "", resp.StatusCode, nil
+		}
+	}
+	return string(body), resp.StatusCode, nil
+}
+
+func (s *Sender) initHttpTransportWithDialer() *http.Client {
+	var dialer net.Dialer
+	tr := &http.Transport{
+		MaxIdleConnsPerHost:   100,
+		ResponseHeaderTimeout: 1000 * time.Millisecond,
+		Proxy:                 http.ProxyFromEnvironment,
+		Dial:                  dialer.Dial,
+	}
+	go func() {
+		for {
+			time.Sleep(10 * time.Second)
+			tr.CloseIdleConnections()
+		}
+	}()
+	client := &http.Client{Transport: tr}
+	client.Timeout = 3000 * time.Millisecond
+	return client
+}
+
 func (s *Sender) Send(event event.Event) {
-	payload := event.Payload()
-	body, err := json.Marshal(payload)
+	// execute http request
+	res, _, err := s.hitEndpoint(event.Context(), event)
```

Comment:
Do we need to relay the status code by to the caller?

Reply:
ok done